### PR TITLE
feat: Basic mfs.write command

### PR DESCRIPTION
**Pull request by achingbrain**
Probably full of bugs.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Warning: This file is automatically synced from https://github.com/ipfs/ci-sync so if you want to change it, please change it there and ask someone to sync all repositories.
 machine:
   node:
-    version: stable
```
Comment (diasdavid) [inline circle.yml]: Please use LTS

Comment (achingbrain) [reply circle.yml]: If we're supposed to be standardising on Jenkins, do you think it might be better to remove the travis & circle setup instead?

Comment (diasdavid) [reply circle.yml]: ✅  from me. cc @VictorBjelkholm

Comment (VictorBjelkholm) [reply circle.yml]: Yeah, code coverage doesn't have a solution as of now. See https://github.com/ipfs/wg-qa-testing-dx/issues/23#issuecomment-386346304

Once we have that in place, we can start removing Travis and Circle from everywhere.

Comment (achingbrain) [reply circle.yml]: Hmm, I changed it to be node version 8 and it's broken the circle build.  Which is odd because the travis build passes with node 8.  Circle complains about leveldown being compiled for a later version of node - do we cache the dependencies between Circle builds?

Comment (VictorBjelkholm) [reply circle.yml]: @achingbrain the caching could be a bit funky in CircleCI somtimes, I've retried the job with cache reset this time. Running here: https://circleci.com/gh/ipfs/js-ipfs-mfs/34

Comment (achingbrain) [reply circle.yml]: Nice one, thanks - that appears to have fixed it.

File: package.json [hunk 2]
```diff
@@ -7,15 +7,16 @@
...
     "test:node": "aegir test -t node",
     "test:browser": "aegir test -t browser -t webworker -f test/browser.js",
     "build": "aegir build",
     "lint": "aegir lint",
     "release": "aegir release",
     "release-minor": "aegir release --type minor",
     "release-major": "aegir release --type major",
-    "coverage": "aegir coverage"
+    "coverage": "aegir coverage",
+    "commit-lint": "commitlint --extends=@commitlint/config-conventional --from=remotes/origin/master --to=HEAD"
```
Comment (VictorBjelkholm) [inline package.json]: Let's not do this here, it'll come from aegir itself

File: package.json [hunk 3]
```diff
@@ -35,37 +36,43 @@
   },
   "homepage": "https://github.com/ipfs/js-ipfs-mfs#readme",
   "devDependencies": {
+    "@commitlint/cli": "^6.2.0",
+    "@commitlint/config-conventional": "^6.1.3",
```
Comment (VictorBjelkholm) [inline package.json]: Same here, dependencies will come from aegir itself

File: package.json (hunk 2 above)
Comment (achingbrain) [reply package.json]: Can I remove it when it's in aegir instead?  Without this it's potluck as to whether Jenkins will quibble about the commit message or not.  It'd be a much better experience to know about it ahead of time.

Comment (VictorBjelkholm) [reply package.json]: Fair enough, sounds reasonable.

**Comment by diasdavid** [comment]
@achingbrain Does this PR implement the full MFS directory tree? What is the ETA to make it compatible with https://github.com/ipfs-shipyard/ipfs-blob-store/blob/master/src/mfs.js ?

**Comment by achingbrain** [comment]
@diasdavid each command is implemented, there are a few edge cases left but it's mostly there.

Not sure on the ETA - probably later this week?  I was going to put this behind an experimental flag for js-ipfs tomorrow and run all the interface tests against it.

Thanks for that link - I notice that `ipfs-blob-store` [expects `ipfs.files.read` to return a ReadableStream](https://github.com/ipfs-shipyard/ipfs-blob-store/blob/master/src/mfs.js#L72) but the spec says [it should return a Buffer](https://github.com/ipfs/interface-ipfs-core/blob/master/SPEC/FILES.md#read) so there will be those sort of inconsistencies to work out.

**Comment by diasdavid** [comment]
@achingbrain sweeeeet! That's awesome news.

> but the spec says it should return a Buffer so there will be those sort of inconsistencies to work out.

That is because we had ReadableStreams for everything when that was developed. Then we started shipping APIs that expose the trio Readable + Pull + Callback/Promise. All of that needs to be updated and propagated, you are correct!

**Comment by diasdavid** [comment]
> Do you mean there are supposed to be 3x methods for everything - read, readReadableStream and readPullStream, each one either taking a callback or returning a promise?

Yes, so it follows the other APIs

**Comment by achingbrain** [comment]
Ok, I've opened ipfs/interface-ipfs-core#265 that adds those signatures to the spec for mfs files.read.
